zigzag/inputs/examples/validation/sram_imc/dimc_validation/28nm/dimc_cost_model.py

```diff
@@ -0,0 +1,304 @@
+        :param number_of_multiplier: The number of multiplier
+        """
+        self.mult = Multiplier(vdd, input_precision, unit_area, unit_delay, unit_cap)
+        self.vdd = vdd
+        self.input_precision = input_precision
+        self.output_precision = input_precision # output precision = input precision
+        self.number_of_multiplier = number_of_multiplier
+    
+    def calculate_area(self):
+        """
+        area: The area cost (unit: mm2)
+        """
+        area = self.mult.calculate_area() * self.number_of_multiplier
+        return area
+        
+    def calculate_delay(self):
+        """
+        delay: The delay cost (unit: ns)
+        """
+        delay = self.mult.calculate_delay()
+        return delay
+        
+    def calculate_energy(self):
+        """
+        energy: The energy cost (unit: fJ)
+        """
+        energy = self.mult.calculate_energy() * self.number_of_multiplier
+        return energy
+
+
+class Adder:
+    def __init__(self, vdd: float, input_precision: int, unit_area: float, unit_delay: float, unit_cap: float):
+        """
+        Class for a {input_precision}-b Carry-Ripple Adder
+        :param vdd:                The supply voltage (unit: V)
+        :param input_precision:    The bit precision of the input (unit: bit)
+        :param output_precision:   The bit precision of the output (unit: bit)
+        :param number_of_1b_adder: The number of 1-b adder in the adder tree
+        """
+        self.nand2 = UnitNand2(unit_area, unit_delay, unit_cap)
+        self.xor2 = UnitXor2(unit_area, unit_delay, unit_cap)
+        self.vdd = vdd
+        self.input_precision = input_precision
+        self.output_precision = input_precision + 1
+        self.number_of_1b_adder = input_precision
+        
+    def calculate_area(self):
+        """
+        area: The area cost (unit: mm2)
+        """
+        area = (3*self.nand2.calculate_area() + 2*self.xor2.calculate_area())*self.number_of_1b_adder
+        return area
+    
+    def calculate_delay_lsb(self):
+        """
+        delay: The delay cost for LSB (unit: ns) (best-case delay, also equals to the delay for Tsum of 1-b adder)
+        """
+        delay_sum = 2*self.xor2.calculate_delay() # 2 XOR gate delay (A-to-Sum)
+        return delay_sum
+    
+    def calculate_delay_msb(self):
+        """
+        delay: The delay cost for MSB (unit: ns) (worst-case delay)
+        """
+        delay_carry = (self.xor2.calculate_delay() + 2*self.nand2.calculate_delay()) + (2*self.nand2.calculate_delay()) * (self.input_precision-1) # A-to-Cout -> Cin-to-Count * (precision-1)
+        return delay_carry
+    
+    def calculate_energy(self):
+        """
+        energy: The energy cost (each time it is triggered) (unit: fJ)
+        """
+        energy = (2*self.xor2.calculate_cap() + 3*self.nand2.calculate_cap()) * self.vdd**2 * self.number_of_1b_adder
+        return energy
+        
+
+class AdderTree:

```

Comment:
It seems to me like these classes (and this file in general) is more general than just for a specific input? If this class is used across multiple instantiations of accelerators, this should be under _classes/hardware/architecture/dimc/..._?

The same question/remark holds for the other validations

Reply:
This is okay. All the classes under validation folder are only used by validation scripts. The correponding classes used within ZigZag will be more complicated, which is under _classes/hardware/architecture/DimcArray.py_